src/core/api/plug_scanners.go

```diff
@@ -0,0 +1,350 @@
+		for _, w := range ws {
+			tw := strings.TrimSpace(w)
+			kv := strings.Split(tw, "=")
+			if len(kv) == 2 {
+				// valid key and value for the querying keyword
+				kws[kv[0]] = kv[1]
+			}
+		}
+	}
+
+	if len(kws) > 0 {
+		query.Keywords = kws
+	}
+
+	all, err := sa.c.ListRegistrations(query)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scanner API: list all"))
+		return
+	}
+
+	// Response to the client
+	sa.Data["json"] = all
+	sa.ServeJSON()
+}
+
+// Create a new scanner
+func (sa *ScannerAPI) Create() {
+	r := &scanner.Registration{}
+
+	if err := sa.DecodeJSONReq(r); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	if err := r.Validate(false); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	// Explicitly check if conflict
+	if !sa.checkDuplicated("name", r.Name) ||
+		!sa.checkDuplicated("url", r.URL) {
+		return
+	}
+
+	uid, err := sa.c.CreateRegistration(r)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scanner API: create"))
+		return
+	}
+
+	location := fmt.Sprintf("%s/%s", sa.Ctx.Request.RequestURI, uid)
+	sa.Ctx.ResponseWriter.Header().Add("Location", location)
+
+	resp := make(map[string]string, 1)
+	resp["uid"] = uid
+
+	// Response to the client
+	sa.Data["json"] = resp
+	sa.ServeJSON()
+}
+
+// Update a scanner
+func (sa *ScannerAPI) Update() {
+	r := sa.get()
+	if r == nil {
+		// meet error
+		return
+	}
+
+	// full dose updated
+	rr := &scanner.Registration{}
+	if err := sa.DecodeJSONReq(rr); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: update"))
+		return
+	}
+
+	if err := r.Validate(true); err != nil {
+		sa.SendBadRequestError(errors.Wrap(err, "scanner API: update"))
+		return
+	}
+
+	// Name changed?
+	if r.Name != rr.Name {
+		if !sa.checkDuplicated("name", rr.Name) {
+			return
+		}
+	}
+
+	// URL changed?
+	if r.URL != rr.URL {
+		if !sa.checkDuplicated("url", rr.URL) {
+			return
+		}
+	}
+
+	getChanges(r, rr)
+
+	if err := sa.c.UpdateRegistration(r); err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scanner API: update"))
+		return
+	}
+
+	location := fmt.Sprintf("%s/%s", sa.Ctx.Request.RequestURI, r.UUID)
+	sa.Ctx.ResponseWriter.Header().Add("Location", location)
+
+	// Response to the client
+	sa.Data["json"] = r
+	sa.ServeJSON()
+}
+
+// Delete the scanner
+func (sa *ScannerAPI) Delete() {
+	uid := sa.GetStringFromPath(":uid")
+	if len(uid) == 0 {
+		sa.SendBadRequestError(errors.New("missing uid"))
+		return
+	}
+
+	deleted, err := sa.c.DeleteRegistration(uid)
+	if err != nil {
+		sa.SendInternalServerError(errors.Wrap(err, "scanner API: delete"))
+		return
+	}
+
+	if deleted == nil {
+		// Not found
+		sa.SendNotFoundError(errors.Errorf("scanner registration: %s", uid))
+		return
+	}
+
+	sa.Data["json"] = deleted
+	sa.ServeJSON()
+}
+
+// SetAsDefault sets the given registration as default one
+func (sa *ScannerAPI) SetAsDefault() {
```

Comment:
Do we need a special API to set the scanner as default? This function can be done by update scanner API.

Reply:
Yes, we need. There is only one default scanner in the system. We need to update the specified one as default and mark the previous default one to not default. More clear.

This method follows the design proposal.

Reply:
But this can be done in update API. When the instance updated and `IsDefault` field changed from false to true, we can update all records' whose `IsDefault` value are true exclude the record currently update and change their `IsDefault` from true to false.

This can be implemented in the manager or DAO of the scanner if it's sure that there is only one default scanner in the system.

Reply:
> But this can be done in update API. When the instance updated and `IsDefault` field changed from false to true, we can update all records' whose `IsDefault` value are true exclude the record currently update and change their `IsDefault` from true to false.
> 
> This can be implemented in the manager or DAO of the scanner if it's sure that there is only one default scanner in the system.

What you described is done in the separate method and API. Meanwhile, this method is proposed by the design proposal and I'm fine with it. So let's keep it as it is now.